atspi-common/src/object_match.rs

```diff
@@ -0,0 +1,281 @@
+#[repr(u32)]
+pub enum TreeTraversalType {
+	/// When traversing the tree, restrict to children of the current object.
+	RestrictChildren,
+
+	/// When traversing the tree, restrict to siblings of the current object.
+	RestrictSibling,
+
+	/// Traverse the tree in order of appearance.
+	#[default]
+	Inorder,
+}
+
+/// Definition of match rules for accessible objects.
+/// Rule(s) against which we can match an  object or a collection thereof.
+///
+/// # Examples
+///  ```Rust
+///     let builder = MatchRule::builder();
+/// ```
+///
+#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
+pub struct ObjectMatchRule {
+	pub states: StateSet,
+	pub states_mt: MatchType,
+	pub attr: HashMap<String, String>,
+	pub attr_mt: MatchType,
+	pub roles: Vec<Role>,
+	pub roles_mt: MatchType,
+	pub ifaces: InterfaceSet,
+	pub ifaces_mt: MatchType,
+	pub invert: bool,
+	// Private phantom, gets compiled away.
+	// Here to ensure the builder is the only route to obtain a `MatchRule`
+	#[serde(skip)]
+	phantom: std::marker::PhantomData<()>,
+}
+
+// !!! WARNING !!! :
+//
+// State and Role are defined as u32 in Accessible.xml but as i32 in Collection.xml
+//
+// The signature on StateSet is defined inconsistently in the XMLs
+// Accessible.xml: GetState type="au"
+// Collection.xml: GetMatches argument name="rule" type="(aiia{ss}iaiiasib)"
+// The latter starts with ai, which is a signature for an array of signed ints i32.
+//
+// https://gitlab.gnome.org/federico/at-spi2-core/-/commit/4885efedeef71e0df8210622771a0b1bb10e194d
+impl Type for ObjectMatchRule {
+	fn signature() -> Signature<'static> {
+		Signature::from_str_unchecked("(aiia{ss}iaiiasib)")
+	}
+}
+
+impl ObjectMatchRule {
+	/// Create a new `MatchRuleBuilder`
+	#[must_use]
+	pub fn builder() -> ObjectMatchRuleBuilder {
+		ObjectMatchRuleBuilder::default()
+	}
+}
+
+/// The 'builder' type for `MatchRule`.  
+/// Use its methods to set match criteria.
+#[derive(Debug, Clone, Default)]
+pub struct ObjectMatchRuleBuilder {
+	states: StateSet,
+	states_mt: MatchType,
+	attr: HashMap<String, String>,
+	attr_mt: MatchType,
+	roles: Vec<Role>,
+	roles_mt: MatchType,
+	ifaces: InterfaceSet,
+	ifaces_mt: MatchType,
+	invert: bool,
+}
+
+impl ObjectMatchRuleBuilder {
+	/// Insert a `StateSet` to the builder
+	#[must_use]
+	pub fn states(mut self, state_set: StateSet, mt: MatchType) -> Self {
+		self.states = state_set;
+		self.states_mt = mt;
+		self
+	}
+
+	/// Insert a map of attributes
+	#[must_use]
+	pub fn attributes(mut self, attributes: HashMap<String, String>, mt: MatchType) -> Self {
+		self.attr = attributes;
+		self.attr_mt = mt;
+		self
+	}
+
+	/// Insert a slice of `Role`s
+	#[must_use]
+	pub fn roles(mut self, roles: &[Role], mt: MatchType) -> Self {
+		self.roles = roles.into();
+		self.roles_mt = mt;
+		self
+	}
+
+	/// Insert a slice of `Interface`s
+	#[must_use]
+	pub fn interfaces(mut self, interfaces: &[Interface], mt: MatchType) -> Self {
```

Comment:
Possibly consider `interfaces: T where T: IntoIteer<Interface>`? I think we've implemented that on `InterfaceSet`.

Reply:
Great idea - thanks!

Reply:
Done!